choreographer/src/main/java/eu/arrowhead/core/choreographer/service/ChoreographerService.java

```diff
@@ -194,46 +215,64 @@ public void receiveSessionStepDoneMessage(ChoreographerSessionRunningStepDataDTO
     }
 
     //-------------------------------------------------------------------------------------------------
-    public void runStep(ChoreographerStep step, long sessionId) throws InterruptedException {
-        logger.debug("Running " + step.getId() + "     " + step.getName() + "       sessionId: " + sessionId + "!");
+    @JmsListener(destination = "session-step-error")
+    public void receiveSessionStepErrorMessage(ChoreographerSessionRunningStepDataDTO sessionRunningStepDataDTO) {
+        //TODO: Implement logic corresponding to what happens when a running step in a session returns from an executor with an error.
+    }
+
+    public void executeStep(ChoreographerStep step, long sessionId) throws InterruptedException {
+        logger.debug("Execution of step with the id of " + step.getId() + " and sessionId of " + sessionId + " started.");
 
         ChoreographerRunningStep runningStep = insertInitiatedRunningStep(step.getId(), sessionId);
 
-        ServiceQueryFormDTO serviceQuery = new ServiceQueryFormDTO();
-        serviceQuery.setServiceDefinitionRequirement(step.getServiceName().toLowerCase());
+        ChoreographerSuitableExecutorResponseDTO suitableExecutors = choreographerDBService.getSuitableExecutorIdsByStepId(step.getId());
 
-        if (sslEnabled) {
-            final PublicKey publicKey = (PublicKey) arrowheadContext.get(CommonConstants.SERVER_PUBLIC_KEY);
-            requesterSystem.setAuthenticationInfo(Base64.getEncoder().encodeToString(publicKey.getEncoded()));
+        if (suitableExecutors.getSuitableExecutorIds().isEmpty()) {
+            //TODO: Research how error handling works in JMS.
         }
 
-        final OrchestrationFormRequestDTO orchestrationForm = new OrchestrationFormRequestDTO.Builder(requesterSystem)
-                                                                                             .requestedService(serviceQuery)
-                                                                                             .flag(OrchestrationFlags.Flag.MATCHMAKING, true)
-                                                                                             .flag(OrchestrationFlags.Flag.OVERRIDE_STORE, true)
-                                                                                             .flag(OrchestrationFlags.Flag.TRIGGER_INTER_CLOUD, false)
-                                                                                             .build();
+        final long executorId = getRandomElementFromList(suitableExecutors.getSuitableExecutorIds());
```

Comment:
move behind an interface

one implementations should check precondition service availability, before selection